Handle escaped quotes

Aflynn50 opened the pull request:
Modified parseStringLiteral to ignore escaped quotes and added testing for it.

fernape commented on internal/parse/parser.go:
```diff
@@ -300,20 +300,22 @@ func (p *Parser) parseInputExpression() (*InputPart, bool, error) {
 
-// parseStringLiteral parses quoted expressions and ignores their content.
+// parseStringLiteral parses quoted expressions and ignores their content
+// including escaped quotes.
```
Maybe add this:
```
+// Returns a pointer to a BypassPart, true and nil on success.
+// Returns nil, false and an error if an error was found.
+// Returns nil, false and nil if no string literal was found.
```

Aflynn50 replied:
I think for changing the comments to this form we could do it all at once in a separate PR so that everything remains  consistent.

solmyrcan commented on internal/parse/parser.go:
```diff
@@ -231,6 +237,39 @@ func (p *Parser) skipName() bool {
+// skipChar jumps over the current char if it matches the byte passed as a
+// parameter. If the char is escaped it is ignored. Returns true in that case,
+// false otherwise.
+func (p *Parser) skipChar(c byte) bool {
```
Maybe we could call `peekByte()` here, like so:

`
if p.peekByte(c) && (!p.isEscaped()) { `

could also do this at the `skipByte()` method, like so:

```
func (p *Parser) skipByte(b byte) bool {
	if p.peekByte(b) { 
```

Aflynn50 replied:
Nice, I've changed skipChar. I'll leave skipByte for now since its not in this PRs remit but its a good point.

manadart commented:
Double quoted strings are not generally used in SQL.

One would prefer 
```sql 
SELECT 'something' FROM dual
```
over
```sql 
SELECT "something" FROM dual
```

And apostrophes are repeated in order to escape. So "Onos T'oolan" would be retrieved thusly:
```sql
SELECT 'Onos T''oolan' FROM dual
```

In turn, selecting a single quote would be:
```sql
SELECT '''' FROM dual
```

Are we covering all of these?

Aflynn50 commented:
> Double quoted strings are not generally used in SQL.
> 
> One would prefer
> 
> ```sql
> SELECT 'something' FROM dual
> ```
> 
> over
> 
> ```sql
> SELECT "something" FROM dual
> ```
> 
> And apostrophes are repeated in order to escape. So "Onos T'oolan" would be retrieved thusly:
> 
> ```sql
> SELECT 'Onos T''oolan' FROM dual
> ```
> 
> In turn, selecting a single quote would be:
> 
> ```sql
> SELECT '''' FROM dual
> ```
> 
> Are we covering all of these?

> Double quoted strings are not generally used in SQL.
> 
> One would prefer
> 
> ```sql
> SELECT 'something' FROM dual
> ```
> 
> over
> 
> ```sql
> SELECT "something" FROM dual
> ```
> 
> And apostrophes are repeated in order to escape. So "Onos T'oolan" would be retrieved thusly:
> 
> ```sql
> SELECT 'Onos T''oolan' FROM dual
> ```
> 
> In turn, selecting a single quote would be:
> 
> ```sql
> SELECT '''' FROM dual
> ```
> 
> Are we covering all of these?

PR changed to use repeated quote escaping rather than backslash escaping.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -305,23 +305,35 @@ func (p *Parser) parseInputExpression() (*InputPart, bool, error) {
-				return nil, false, fmt.Errorf("missing right quote in string literal")
+			mark := p.pos
+
+			if p.skipByte(c) {
```
Isn't all of the above logic, starting at the very first `if` block, a manual version of `if p.skipByte('"') || p.skipByte('\'')`?

Aflynn50 replied:
You're right, I think I misunderstood the flow of the code when we talked in the call. I've changed it.

niemeyer reviewed (changes requested):
Just one comment about the implementation. Idea looks nice again.

Aflynn50 committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

Added handeling for escaped strings

# This is the commit message #2:

Added tests for escaped string handeling

Aflynn50 committed:
# This is a combination of 4 commits.
# This is the 1st commit message:

Added handeling for escaped strings

# This is the commit message #2:

Added tests for escaped string handeling

# This is the commit message #3:

Deal with case where the backslash before the quote is escaped

# This is the commit message #4:

Add position to error thrown by parseStringLiteral

Aflynn50 committed:
# This is a combination of 6 commits.
# This is the 1st commit message:

Added handeling for escaped strings

# This is the commit message #2:

Added tests for escaped string handeling

# This is the commit message #3:

Deal with case where the backslash before the quote is escaped

# This is the commit message #4:

Add position to error thrown by parseStringLiteral

# This is the commit message #5:

Change relevent tests to table driven format

# This is the commit message #6:

Change skipChar to use peekByte

Aflynn50 committed:
Added handeling for escaped strings

Added tests for escaped string handeling

Deal with case where the backslash before the quote is escaped

Add position to error thrown by parseStringLiteral

Change relevent tests to table driven format

Change skipChar to use peekByte

Changed escaping to work with repeated quotes rather than backslash

Aflynn50 committed:
Added handeling for escaped strings

Added tests for escaped string handeling

Deal with case where the backslash before the quote is escaped

Add position to error thrown by parseStringLiteral

Change relevent tests to table driven format

Change skipChar to use peekByte

Changed escaping to work with repeated quotes rather than backslash

Add more tests for quote matching. (#9)

* Added handeling for escaped strings

* Added tests for escaped string handeling

* Deal with case where the backslash before the quote is escaped

* Add position to error thrown by parseStringLiteral

* Change relevent tests to table driven format

* Change skipChar to use peekByte

* Changed escaping to work with repeated quotes rather than backslash

* Add more tests for quote matching.

* Undo change made by mistake.

Co-authored-by: Alastair Flynn <alastairflynn@hotmail.co.uk>

Change manuel version of skipByte to disjunction of skipByte

Aflynn50 committed:
Merged with refactored code

niemeyer commented on internal/expr/parser.go:
```diff
@@ -305,23 +305,29 @@ func (p *Parser) parseInputExpression() (*inputPart, bool, error) {
-	}
 
+		// Reached end of string and didn't find the closing quote
+		return nil, false, fmt.Errorf("missing right quote for char %d in string literal", cp.pos)
```
Thanks for the changes. Only one trivial: can we please update this to use the same `column %d:` prefix style we discussed elsewhere?

Aflynn50 replied:
Changed.

Aflynn50 committed:
Change error to use column instead

Aflynn50 committed:
Replace mark with cp.pos
